valgrind_suppressions.txt

```diff
@@ -416,3 +416,20 @@
    fun:_ZN16cppmicroservices7scrimpl19SCRAsyncWorkService4postEOSt13packaged_taskIFvvEE
    ...
 }
+
+{
+   <usDeclarativeServicesTestsSuppression2>
+   Memcheck:Value8
+   ...
+   fun:_ZN7testing8internal20PrintBytesInObjectToEPKhmPSo
+   ...
+}
+
+{
+   <usDeclarativeServicesTestsSuppression3>
+   Memcheck:Cond
+   ...
+   fun:_ZN7testing8internal20PrintBytesInObjectToEPKhmPSo
+   ...
```

Comment:
which tests were these valgrind errors coming from?
If I remember correctly, these come from not having an overloaded << operator for a user-defined type.

Reply:
The errors that these suppress come from gtest/gmock's warmup for `usDeclarativeServicesTests`. Looks like it has something to do with `ComponentMetadata`? Should I declare an overloaded << operator for it? Additionally the error only shows up a couple times in the log but the error summary for valgrind states there are 2448 errors in total; including the suppression above removes them all (just for context).

```
==32383== Memcheck, a memory error detector
==32383== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==32383== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==32383== Command: ./build/bin/usDeclarativeServicesTests
==32383==
==32383== Use of uninitialised value of size 8
==32383==    at 0x7471911: _itoa_word (_itoa.c:180)
==32383==    by 0x747500D: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C4D2D: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7471918: _itoa_word (_itoa.c:180)
==32383==    by 0x747500D: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C4D2D: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7475114: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C4D2D: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7475C4C: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C4D2D: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7475114: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C76C1: testing::(anonymous namespace)::PrintByteSegmentInObjectTo(unsigned char const*, unsigned long, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64C4D91: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7475C4C: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C76C1: testing::(anonymous namespace)::PrintByteSegmentInObjectTo(unsigned char const*, unsigned long, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64C4D91: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
[==========] Running 323 tests from 36 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from ActivatorTest
[ RUN      ] ActivatorTest.VerifySCRService
```